### PR TITLE
(Heterogeneous) `NeighborLoader`

### PR DESCRIPTION
**Pull request by rusty1s**
Screen Shot 2021-08-11 at 12 58 25

### CODE REVIEW

File: torch_geometric/loader/neighbor_loader.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+        **kwargs (optional): Additional arguments of
+            :class:`torch.utils.data.DataLoader`, such as :obj:`batch_size`,
+            :obj:`shuffle`, :obj:`drop_last` or :obj:`num_workers`.
+    """
+    def __init__(
+        self,
+        data: Union[Data, HeteroData],
+        num_neighbors: Union[List[int], Dict[EdgeType, List[int]]],
+        input_nodes: Union[OptTensor, NodeType, Tuple[NodeType,
+                                                      OptTensor]] = None,
```
Comment (mrjel) [inline torch_geometric/loader/neighbor_loader.py:125]: Why is a dictionary not allowed for the input_nodes parameter? What if I want to specify a mask for and sample from all node types?

Comment (rusty1s) [reply torch_geometric/loader/neighbor_loader.py:125]: Yeah, the C++ sampling part supports that, but it's a bit tricky to sample input node indices across different node types using PyTorch. Currently, this is implemented as follows:
```python
super().__init__(self.input_nodes[1].tolist(), ...)
```
I don't currently see an elegant way to move this to a dictionary of inputs :( (@RexYing, @JiaxuanYou pointers appreciated)

Comment (rusty1s) [reply torch_geometric/loader/neighbor_loader.py:125]: One option is to convert a dictionary of `input_nodes` into a list holding tuples:
```
[('paper', 0), ('paper', 1), ('author', 1000), ...]
```
and sample from there, but this will create some additional overhead.

File: torch_geometric/loader/neighbor_loader.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+                             **kwargs)
+
+        else:  # `HeteroData`
+            self.node_types, self.edge_types = data.metadata()
+            self.sample_fn = torch.ops.torch_sparse.hetero_neighbor_sample
+            out = to_hetero_csc(data)
+            self.colptr_dict, self.row_dict, self.perm_dict = out
+            if isinstance(num_neighbors, (list, tuple)):
+                self.num_neighbors = {
+                    edge_type_to_str(edge_type): num_neighbors
```
Comment (chantat) [inline torch_geometric/loader/neighbor_loader.py:163]: knit: makes more sense to add the following comment here:
```
    # C++/CUDA functions cannot take dictionaries with tuples as key as input.
    # We therefore need to convert the edge type triplet into a single string.
```

File: torch_geometric/loader/neighbor_loader.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+        node, row, col, edge = self.sample_fn(
+            self.colptr,
+            self.row,
+            torch.tensor(indices),
+            self.num_neighbors,
+            self.replace,
+            self.directed,
+        )
+
+        data = filter_data(self.data, node, row, col, edge, self.perm)
```
Comment (chantat) [inline torch_geometric/loader/neighbor_loader.py:191]: What does filter_data do again?

File: torch_geometric/loader/utils.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+import copy
+
+import torch
+from torch import Tensor
+from torch_sparse import SparseTensor
+
+from torch_geometric.data import Data, HeteroData
+from torch_geometric.data.storage import NodeStorage, EdgeStorage
+
```
Comment (chantat) [inline torch_geometric/loader/utils.py:12]: Need a bit more documentation for these public functions.

**Review by chantat — APPROVED** [review]

**Comment by rusty1s** [comment]
Thanks @chantat for the review. I added more documentation according to your suggestions.

File: torch_geometric/loader/neighbor_loader.py (hunk 1 above)
Comment (mrjel) [reply torch_geometric/loader/neighbor_loader.py:125]: I think I misunderstood the purpose of the input_nodes parameter. I thought you provide nodes indices for nodes from which you sample, instead of node indices of nodes for which you sample neighborhoods.

Yes, I guess in this case this feature would not be that useful anyway and is probably not worth the effort.

File: torch_geometric/loader/neighbor_loader.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+            In heterogeneous graphs, may also take in a dictionary denoting
+            the amount of neighbors to sample for each individual edge type.
+        input_nodes (torch.Tensor or str or Tuple[str, torch.Tensor]): The
+            indices of nodes that should be considered for creating
+            mini-batches.
+            Needs to be either given as a :obj:`torch.LongTensor` or
+            :obj:`torch.BoolTensor`.
+            If set to :obj:`None`, all nodes will be considered.
+            In heterogeneous graphs, needs to be passed as a tuple, denoting
+            the node type and node indices. (default: :obj:`None`)
```
Comment (mrjel) [inline torch_geometric/loader/neighbor_loader.py:108]: Could maybe be made more specific:
"The indices of nodes that should be considered for creating mini-batches."
->
"The indices of nodes for which neighbors are sampled to create mini-batches."

**Review by mrjel — APPROVED** [review]
